src/rdiffbackup/locations/_repo_shadow.py

```diff
@@ -1048,6 +1060,88 @@ def init_owners_mapping(cls, users_map, groups_map, preserve_num_ids):
         map_owners.init_users_mapping(users_map, preserve_num_ids)
         map_owners.init_groups_mapping(groups_map, preserve_num_ids)
 
+# ### LOCKING ####
+
+    # @API(RepoShadow.is_locked, 201)
+    @classmethod
+    def is_locked(cls, lockfile, exclusive):
+        """
+        Validate if the repository is locked or not by the file
+        'rdiff-backup-data/lock.yml'
+
+        Returns True if the file exists and is locked, else returns False
+        """
+        # we need to make sure we have the last state of the lock
+        lockfile.setdata()
+        if not lockfile.lstat():
+            return False  # if the file doesn't exist, it can't be locked
+        with open(lockfile, cls.LOCK_MODE[exclusive]["open"]) as lockfd:
+            try:
+                locking.lock(lockfd, cls.LOCK_MODE[exclusive]["lock"])
+                return False
+            except BlockingIOError:
+                return True
+
+    # @API(RepoShadow.lock, 201)
+    @classmethod
+    def lock(cls, lockfile, exclusive, force=False):
+        """
+        Write a specific file 'rdiff-backup-data/lock.yml' to grab the lock,
+        and verify that no other process took the lock by comparing its
+        content.
+
+        Return True if the lock could be taken, False else.
+        """
+        if cls._lockfd:  # we already opened the lockfile
+            return False
+        pid = os.getpid()
+        identifier = {
+            'timestamp': Globals.current_time_string,
+            'pid': pid,
+            'cmd': psutil.get_pid_name(pid),
+            'hostname': socket.gethostname(),
+        }
+        id_yaml = yaml.safe_dump(identifier)
+        lockfile.setdata()
+        if not lockfile.lstat():
+            if exclusive:
+                open_mode = cls.LOCK_MODE[exclusive]["truncate"]
+            else:
+                # we can't take the lock if the file doesn't exist
+                return False
+        else:
+            open_mode = cls.LOCK_MODE[exclusive]["open"]
+        try:
+            lockfd = open(lockfile, open_mode)
+            locking.lock(lockfd, cls.LOCK_MODE[exclusive]["lock"])
+            if exclusive:  # let's keep a trace of who's writing
+                lockfd.seek(0)
+                lockfd.truncate()
+                lockfd.write(id_yaml)
+                lockfd.flush()
+            cls._lockfd = lockfd
+            return True
+        except BlockingIOError:
+            if lockfd:
+                lockfd.close()
+            return False
+
+    # @API(RepoShadow.unlock, 201)
+    @classmethod
+    def unlock(cls, lockfile, exclusive):
+        """
+        Remove any lock existing.
+
+        We don't check for any content because we have the lock and should be
+        the only process running on this repository.
+        """
+        if cls._lockfd:
```

Comment:
Shoudn't we call `locking.unlock()` somewhere ?

Reply:
Closing the file releases the lock automatically.